mkosi

```diff
@@ -1228,6 +1255,58 @@ def patch_root_uuid(args, loopdev, root_hash):
         subprocess.run(["sfdisk", "--part-uuid", loopdev, str(args.root_partno), str(u)],
                        check=True)
 
+def install_unified_kernel(args, workspace, run_build_script, root_hash):
+
+    # Iterates through all kernel versions included in the image and
+    # generates a combined kernel+initrd+cmdline+osrelease EFI file
+    # from it and places it in the /EFI/Linux directory of the
+    # ESP. sd-boot iterates through them and shows them in the
+    # menu. These "unified" single-file images have the benefit that
+    # they can be signed like normal EFI binaries, and can encode
+    # everything necessary to boot a specific root device, including
+    # the root hash.
+
+    if not args.bootable:
+        return
+
+    if args.distribution != Distribution.fedora:
+        return
+
+    with complete_step("Generating combined kernel + initrd boot file"):
+
+        cmdline = args.kernel_commandline
+        if root_hash is not None:
+            cmdline += " roothash=" + root_hash
+
+            for kver in os.scandir(os.path.join(workspace, "root", "usr/lib/modules")):
+                if not kver.is_dir():
+                    continue
+
+                boot_binary = "/efi/EFI/Linux/linux-" + kver.name
+                if root_hash is not None:
+                    boot_binary += "-" + root_hash
+                boot_binary += ".efi"
+
+                dracut = ["/usr/bin/dracut",
+                          "-v",
+                          "--no-hostonly",
+                          "--uefi",
+                          "--kver", kver.name,
+                          "--kernel-cmdline", cmdline ]
+
+                # Temporary fix until dracut includes these in the image anyway
+                dracut += [ "-i", "/usr/lib/systemd/system/systemd-volatile-root.service", "/usr/lib/systemd/system/systemd-volatile-root.service",
+                            "-i", "/usr/lib/systemd/systemd-volatile-root", "/usr/lib/systemd/systemd-volatile-root",
+                            "-i", "/usr/lib/systemd/systemd-veritysetup", "/usr/lib/systemd/systemd-veritysetup",
+                            "-i", "/usr/lib/systemd/system-generators/systemd-veritysetup-generator", "/usr/lib/systemd/system-generators/systemd-veritysetup-generator" ]
```

Comment:
well, it's mostly a hack, and I kinda hope this can go away quickly, when dracut gets updated to include these things in the "no-host-only" mode anyway. Hence, I wanted it to look a bit like a hack ;-)